clippy_lints/src/methods/collapsible_str_replace.rs

```diff
@@ -0,0 +1,160 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::get_parent_expr;
+use clippy_utils::visitors::for_each_expr;
+use clippy_utils::SpanlessEq;
+use core::ops::ControlFlow;
+use if_chain::if_chain;
+use rustc_ast::ast::LitKind;
+use rustc_errors::Applicability;
+use rustc_hir as hir;
+use rustc_hir::{ExprKind, Path, QPath};
+use rustc_lint::LateContext;
+use rustc_span::source_map::Spanned;
+use rustc_span::Span;
+
+use super::method_call;
+use super::COLLAPSIBLE_STR_REPLACE;
+
+pub(super) fn check<'tcx>(cx: &LateContext<'tcx>, expr: &'tcx hir::Expr<'tcx>, to: &'tcx hir::Expr<'tcx>) {
+    if_chain! {
+        if let Some(replace_methods) = collect_replace_calls(cx, expr, to);
+        if let Some(earliest_replace_call) = replace_methods.methods.last();
+        if let Some((_, [earliest_recv, ..], _)) = method_call(earliest_replace_call);
+        if cx.typeck_results().expr_ty(earliest_recv).peel_refs().is_str();
+
+        then {
+            // If the parent node's `to` argument is the same as the `to` argument
+            // of the last replace call in the current chain, don't lint as it would be duplicated lint
+            if let Some(parent) = get_parent_expr(cx, expr)
+                && let Some((name, [_, _, current_to], _)) = method_call(parent)
+                && name == "replace"
+                && SpanlessEq::new(cx).eq_expr(to, current_to)
+            {
+                return;
+            }
+
+            // Check if there's an earlier `str::replace` call
+            if let Some((_, [..], span_lo)) = method_call(earliest_replace_call) {
+                check_consecutive_replace_calls(cx, expr, &replace_methods, to, span_lo);
+            }
+        }
+    }
+}
+
+struct ReplaceMethods<'tcx> {
+    methods: Vec<&'tcx hir::Expr<'tcx>>,
+    args: Vec<&'tcx hir::Expr<'tcx>>,
+}
+
+fn collect_replace_calls<'tcx>(
+    cx: &LateContext<'tcx>,
+    expr: &'tcx hir::Expr<'tcx>,
+    to_arg: &'tcx hir::Expr<'tcx>,
+) -> Option<ReplaceMethods<'tcx>> {
+    let mut methods = Vec::new();
+    let mut from_args = Vec::new();
+    let mut return_none = false;
+
+    let _: Option<()> = for_each_expr(expr, |e| {
+        if let Some((name, [_, args @ ..], _)) = method_call(e) {
+            match (name, args) {
+                ("replace", [from, to]) => {
+                    if SpanlessEq::new(cx).eq_expr(to_arg, to)
+                        && cx.typeck_results().expr_ty(from).peel_refs().is_char()
```

Comment:
You were right. This should not only check for `char` but also strings, and `paths`. So pretty much the same you do later when formatting the `from_args`. It should probably share the same code. So I would add a function

```rust
fn is_valid_from_arg(...) {
    matches!(expr.kind, Str | Char | Path)
}
```

And then simplify the `get_replace_call_char_arg_repr` function to

```rust
snippet(from_arg, "_").replace_start("'", '"').replace_end("'", '"')
```

You don't even have to check for the `ExprKind` then, because if it would be of a wrong `ExprKind` the `from_arg` would've never been added to the `from_args` list.

Reply:
> You don't even have to check for the ExprKind then, because if it would be of a wrong ExprKind the from_arg would've never been added to the from_args list.
You're right here. The `is_valid_from_arg` is also not necessary since `cx.typeck_results().expr_ty(from).peel_refs().is_char()` checks if the underlying value of a `Path` is a `char` and we don't check for `str` in this iteration. So in the end we only need `snippet`.